src/Storage/Controllers/DataLockController.cs

```diff
@@ -0,0 +1,142 @@
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.Cosmos;
+
+namespace Altinn.Platform.Storage.Controllers;
+
+/// <summary>
+/// API for handling locking and unlocking of data elements
+/// </summary>
+[Route("storage/api/v1/instances/{instanceOwnerPartyId:int}/{instanceGuid:guid}/data/{dataGuid:guid}/lock")]
+[ApiController]
+public class DataLockController : ControllerBase
+{
+    private readonly IInstanceRepository _instanceRepository;
+    private readonly IDataRepository _dataRepository;
+    private readonly IAuthorization _authorizationService;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DataLockController"/> class
+    /// </summary>
+    /// <param name="instanceRepository">the instance repository</param>
+    /// <param name="dataRepository">the data repository handler</param>
+    /// <param name="authorizationService">the authorization service.</param>
+    public DataLockController(
+        IInstanceRepository instanceRepository,
+        IDataRepository dataRepository,
+        IAuthorization authorizationService)
+    {
+        _instanceRepository = instanceRepository;
+        _dataRepository = dataRepository;
+        _authorizationService = authorizationService;
+    }
+
+    /// <summary>
+    /// Locks a data element
+    /// </summary>
+    /// <param name="instanceOwnerPartyId">The party id of the instance owner.</param>
+    /// <param name="instanceGuid">The id of the instance that the data element is associated with.</param>
+    /// <param name="dataGuid">The id of the data element to delete.</param>
+    /// <returns></returns>
+    [Authorize(Policy = AuthzConstants.POLICY_INSTANCE_WRITE)]
+    [HttpPut]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status201Created)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [Produces("application/json")]
+    public async Task<ActionResult<DataElement>> Lock(int instanceOwnerPartyId, Guid instanceGuid, Guid dataGuid)
+    {
+        (Instance? instance, ActionResult? instanceError) = await GetInstanceAsync(instanceGuid, instanceOwnerPartyId);
+        if (instance == null)
+        {
+            return instanceError!;
+        }
+        
+        DataElement? dataElement = instance.Data.Find(d => d.Id == dataGuid.ToString());
+        
+        if (dataElement?.Locked is true)
+        {
+            return Ok(dataElement);
+        }
+
+        Dictionary<string, object> propertyList = new()
+        {
+            { "/locked", true }
+        };
+        
+        try
+        {
+            DataElement updatedDataElement = await _dataRepository.Update(instanceGuid, dataGuid, propertyList);
+            return Created(updatedDataElement.Id, updatedDataElement);
+        }
+        catch (CosmosException e)
```

Comment:
We're in the process of changing from Cosmos to PostgreSQL. It's important that nothing related to Cosmos appear outside the Repository classes.